### ISSUE TITLE
Support autofixes from TSLint 4.0 that consists of > 1 replacements

### ISSUE THREAD

**Issue by rlasjunies** [Microsoft/vscode-tslint#135]
Hi @egamma 

 TSLint 4 provide the fixes like this:
![image](https://cloud.githubusercontent.com/assets/1439791/20599108/bb0229be-b24c-11e6-9484-ee4faf3b6cc9.png)

 Important information for us are: 
* startPosition
* endPosition
* innerReplacements[]

start/endPosition define where is the problem and innerReplacement are the information to apply the fix. 

This were is the bug. **The code apply the fix based on the start/endPosition. It should not**

I think we should apply the fix based on the information from innerReplacement. 
export interface TSLintFixReplacement {
	innerLength: number;
	innerStart: number;
	innerText: string;
}

My problem to go on is: **how to convert start + lenght in vscode Position structure?**

The edit.replace method of vscode server use a Range of Position.
```
export interface Position {
    /**
     * Line position in a document (zero-based).
     */
    line: number;
    /**
     * Character offset on a line in a document (zero-based).
     */
    character: 
``` 

Is there a way to convert "absolute" position (from innerReplacements) to vscodePosition structure?

**Comment by egamma** [comment]
@rlasjunies the TextDocument provides the method `positionAt`. It converts an offset to a position.

**Comment by egamma** [comment]
PR https://github.com/Microsoft/vscode-tslint/pull/138 fixes the cases where a fix consists of a single replacement. Keeping this issue open to track the support for multiple replacements.

**Comment by egamma** [comment]
@rlasjunies just a quick update. I've refactored the code so that it uses the TSLint provided `RuleFailure` type and its satellite types like `Replacement`. In this way we no longer need to define custom types for auto fixing. Also when use the TSLint provide types then replacements make more sense.

I suggest that the custom vscFix also use the TSLint types like Replacement. I haven't converted the existing vscFixes yet and have disabled them.

**Comment by rlasjunies** [comment]
@egamma your choice is the good one, I did not see `RuleFailure` before this is a great opportunity to reduce incompatilities in the future.

About the vscFix, I've push it a bit but if you think is not needed you may remove it. If the need really appears we may decide to put it back. 

And ... great job in the last days to support the new version tslint 👍

**Comment by egamma** [comment]
@rlasjunies thank you for noticing, that the tslint 4.0 adoption was a swirl... it was an interesting exercise of breaking an API without deprecating the old one and supporting (version -1 compatibility)... I hope the tslint team does some retro on this 😄 

I've just started to add support for multiple Replacements in a Fix. Multiple replacements are already supported by the client request, so this just needs some refactoring on the server. It is great to build on your pioneering work. 

Regarding `vscFix` I suggest to let these fixes sit for a while and see whether we get feedback to revive them.

**Comment by egamma** [comment]
Fixed and published as version 0.6.6 in the marketplace.

@rlasjunies as mentioned above I've removed the vsc fixes for now.

**Comment by egamma** [comment]
@rlasjunies not having the custom fixes confused users (see #144, #145, #136). So I've revived most of the custom quick fixes. I still do not want to provide quick fixes for tslint warnings for which we have editor support, eg., trailing white space, trailing newline.

**Comment by rlasjunies** [comment]
Ok, no prob.
Let me know if you would me to help somewhere.
